wolfcrypt/src/random.c

```diff
@@ -800,6 +809,779 @@ int wc_RNG_TestSeed(const byte* seed, word32 seedSz)
+}
+#elif !defined(ENTROPY_MEMUSE_THREAD) && \
+      (defined(__APPLE__) || defined(__MACH__))
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit time in nanoseconds.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    return mach_absolute_time();
+}
+#elif !defined(ENTROPY_MEMUSE_THREAD) && defined(__aarch64__)
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit timer count.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    word64 cnt;
+    __asm__ __volatile__ (
+        "mrs %[cnt], cntvct_el0"
+        : [cnt] "=r"(cnt)
+        :
+        :
+    );
+    return cnt;
+}
+#elif !defined(ENTROPY_MEMUSE_THREAD) && (_POSIX_C_SOURCE >= 199309L)
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit time that is the nanoseconds of current time.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    struct timespec now;
+
+    clock_gettime(CLOCK_REALTIME, &now);
+
+    return now.tv_nsec;
+}
+#elif !defined(SINGLE_THREADED) && defined(_POSIX_THREADS) && \
+      !defined(__MINGW32__)
+
+/* Start and stop thread that counts as a proxy for time counter. */
+#define ENTROPY_MEMUSE_THREADED
+
+/* Data for entropy thread. */
+typedef struct ENTROPY_THREAD_DATA {
+    /* Current counter - proxy for time. */
+    word64 counter;
+    /* Whether to stop thread. */
+    int stop;
+} ENTROPY_THREAD_DATA;
+
+/* Track whether entropy thread has been started already. */
+static int entropy_thread_started = 0;
+/* Cache thread id for joining on exit. */
+static THREAD_TYPE entropy_thread_id = 0;
+/* Data for thread to update/observer. */
+static volatile ENTROPY_THREAD_DATA entropy_thread_data = { 0, 0 };
+
+/* Get the high resolution time counter. Counter incremented in thread.
+ *
+ * @return  64-bit counter.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    /* Return counter update in thread. */
+    return entropy_thread_data.counter;
+}
+
+/* Thread that increments counter while not told to stop.
+ *
+ * @param [in,out] args  Entropy data including: counter and stop flag.
+ * @return  NULL always.
+ */
+static THREAD_RETURN WOLFSSL_THREAD Entropy_IncCounter(void* args)
+{
+    (void)args;
+
+    /* Thread resources to be disposed of. */
+    pthread_detach(pthread_self());
+
+    /* Keep going until caller tells us to stop and exit. */
+    while (!entropy_thread_data.stop) {
+        /* Increment counter acting as high resolution timer. */
+        entropy_thread_data.counter++;
+    }
+
+#ifdef WOLFSSL_DEBUG_ENTROPY_MEMUSE
+    fprintf(stderr, "EXITING ENTROPY COUNTER THREAD\n");
+#endif
+    /* Exit from thread. */
+    pthread_exit(NULL);
+}
+
+/* Start a thread that increments counter if not one already.
+ *
+ * Won't start a new thread if one already running.
+ * Waits for thread to start by waiting for counter to have incremented.
+ *
+ * @return  0 on success.
+ * @return  Negative on failure.
+ */
+static int Entropy_StartThread()
```

Comment:
`static int Entropy_StartThread (void)`

Reply:
Changed all static functions without parameters to have void.